File: JavaScript/Trie/L_replaceWord.js
class TrieNode {
    constructor() {
        this.children = {};
        this.isEnd = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }

    insert(word) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
        }
        node.isEnd = true;
    }

    search(word) {
        let node = this.root;
        let prefix = "";
        for (const char of word) {
            if (!node.children[char]) {
                return word;
            }
            prefix += char;
            if (node.children[char].isEnd) {
                return prefix;
            }
            node = node.children[char];
        }
        return word;
    }
}

var replaceWords = function (dictionary, sentence) {
    const trie = new Trie();
    for (const word of dictionary) {
        trie.insert(word);
    }

    const words = sentence.split(" ");
    const result = words.map((word) => trie.search(word));

    return result.join(" ");
};

File: Python/Tries/L_replaceWord.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.isEnd = True
    
    def search(self, word):
        node = self.root
        prefix = ''
        for char in word:
            if char not in node.children:
                return word
            prefix += char
            if node.children[char].isEnd:
                return prefix
            node = node.children[char]
        return word

class Solution:
    def replaceWords(self, dictionary, sentence):
        trie = Trie()
        for word in dictionary:
            trie.insert(word)
        
        result = []
        for word in sentence.split():
            result.append(trie.search(word))
        
        return ' '.join(result)